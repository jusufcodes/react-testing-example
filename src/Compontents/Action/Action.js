import React, {useState} from "react";
import ShareIcon from "../Icon/Share.svg";
import FavoriteIcon from "../Icon/Favorite.svg";
import classNames from "classnames";

export const Action = (props) => {
  const [clickCount, setClickCount] = useState(0);

  const handleClick = (event) => {
    setClickCount(clickCount + 1);

    if (props.onClick) {
      props.onClick(event);
    }
  };

  return props.children ? (
    <button onClick={handleClick}>
      {props.children} ({clickCount})
    </button>
  ) : null;
};

export async function likeAndShare(postId) {
  const apiUrl = postId ? "posts" : "error";

  const response = await fetch(`https://jsonplaceholder.typicode.com/${apiUrl}`, {
    method: 'POST',
    body: JSON.stringify({
      id: postId
    }),
    headers: {
      "Content-type": "application/json; charset=UTF-8"
    }
  });

  const json = await response.json();

  return json.id;
}

export const FavoriteAction = (props) => {
  const [liked, setLike] = useState(false);

  async function handleClick() {
    const currentLiked = liked;

    setLike(!currentLiked);

    if (!props.postId) {
      return
    }

    const id = await likeAndShare(props.postId);

    if (typeof id === "number") {
      setLike(!currentLiked);
    } else {
      alert("error");
    }
  }

  return props.postId ? (
    <Action onClick={handleClick}>
      <img src={FavoriteIcon} alt="" className={classNames("favoriteIcon", {
        active: liked
      })}/>
    </Action>
  ) : null;
};

export const ShareAction = (props) => {
  const [shared, setShared] = useState(false);

  async function handleClick() {
    const currentShared = shared;

    setShared(!currentShared);

    if (!props.postId) {
      return
    }

    const id = await likeAndShare(props.postId);

    if (typeof id === "number") {
      setShared(!currentShared);
    } else {
      alert("error");
    }
  }

  return props.postId ? (
    <Action onClick={handleClick}>
      <img src={ShareIcon} alt="" className={classNames("shareIcon", {
        active: shared
      })}/>
    </Action>
  ) : null;
};