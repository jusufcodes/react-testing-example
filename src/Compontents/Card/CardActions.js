import React from "react";

export const CardActions = (props) => {
  return (props.left || props.right) ? (
    <div className={"cardActions"}>
      {props.left && (
        <div className={"cardActionsLeft"}>
          {props.left}
        </div>
      )}
      {props.right && (
        <div className={"cardActionsRight"}>
          {props.right}
        </div>
      )}
    </div>
  ) : null;
};