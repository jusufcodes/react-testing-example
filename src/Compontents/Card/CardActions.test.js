import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import {CardActions} from "./CardActions";
import {Action, FavoriteAction} from "../Action/Action";

describe("CardActions Component", () => {

  test("CardActions without content", () => {
    const actions = <CardActions />;

    expect(shallow(actions).isEmptyRender()).toEqual(true);
  });

  test("CardActions with left content", () => {
    const actionsLeft = <CardActions left={(<Action>Action 1</Action>)} />;

    expect(shallow(actionsLeft).isEmptyRender()).toEqual(false);
    expect(shallow(actionsLeft).children()).toHaveLength(1);

    const actionSnaphot = renderer.create(actionsLeft).toJSON();
    expect(actionSnaphot).toMatchSnapshot();
  });

  test("CardActions with right content", () => {
    const actionsRight = <CardActions right={(<FavoriteAction />)} />;

    expect(shallow(actionsRight).isEmptyRender()).toEqual(false);
    expect(shallow(actionsRight).children()).toHaveLength(1);

    const actionSnaphot = renderer.create(actionsRight).toJSON();
    expect(actionSnaphot).toMatchSnapshot();
  });

  test("CardActions with left and right content", () => {
    const actions = <CardActions
      left={(<Action>Action 1</Action>)}
      right={(<FavoriteAction />)}
    />;

    expect(shallow(actions).isEmptyRender()).toEqual(false);
    expect(shallow(actions).children()).toHaveLength(2);

    const actionSnaphot = renderer.create(actions).toJSON();
    expect(actionSnaphot).toMatchSnapshot();
  });
});