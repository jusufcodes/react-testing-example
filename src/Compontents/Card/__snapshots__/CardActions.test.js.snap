// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardActions Component CardActions with left and right content 1`] = `
<div
  className="cardActions"
>
  <div
    className="cardActionsLeft"
  >
    <button
      onClick={[Function]}
    >
      Action 1
       (
      0
      )
    </button>
  </div>
  <div
    className="cardActionsRight"
  />
</div>
`;

exports[`CardActions Component CardActions with left content 1`] = `
<div
  className="cardActions"
>
  <div
    className="cardActionsLeft"
  >
    <button
      onClick={[Function]}
    >
      Action 1
       (
      0
      )
    </button>
  </div>
</div>
`;

exports[`CardActions Component CardActions with right content 1`] = `
<div
  className="cardActions"
>
  <div
    className="cardActionsRight"
  />
</div>
`;
